### DISCUSSION TITLE
invalid port number MSSQL

### DISCUSSION THREAD

**Discussion by SergKS8** [sfu-db/connector-x#319] [Q&A] [answered]
Hi there!
The problem is that 
when I'm trying to read data from MSSQL like this:
conn = f'mssql://{login}:{password}@{host}/{database}'
cx.read_sql(conn, sql_aims)

I'm getting the error 
**RuntimeError: parse error: invalid port number.**
But default port = 1433 is a valid port.

But if I changed the password in conn string to combination without symbol '#' I get another arror - 
**RuntimeError: Timed out in bb8.**
I think that thing is that my password contain specific symbol # and this is the cфuse of the error.
How can I read data from MSSQL with connectorx in such a case?

The SOURSE - MS SQL Server with SQL Server Authentication.

So, please help me.

**Comment by wangxiaoying** [comment, marked as answer]
Hi @SergKS8 , in case of the special symbols, you can use urlencoding to encode your password in connection string. Here is an example:

```
import urllib.parse
f"mssql://{username}:{urllib.parse.quote_plus(password)}@{host}:{port}/{database}"
```

As for the port, can you try to explicitly specify your port in the connection string?

**Reply by SergKS8** [reply]
@wangxiaoying,  urllib.parse is working well. I've managed to read data from MSSQL using connectorx. Thank you for the soon answer. Have a good day!

**Reply by evbo** [reply]
@wangxiaoying I'm wondering if all kinds of special characters are supported even if they are urlencoded?

If I use pymssql, I can connect just fine (so I know this user works):
```python
import pymssql
conn = pymssql.connect("fehchdev.com", "FEHCH\oper_dev", "b3k@!!1aa", "registrar_home")
cursor = conn.cursor(as_dict=True)
cursor.execute("select * from dbs1.users_home")
row = cursor.fetchone()
print(row)
```

If I use connectorx, I get a `[tiberius](tiberius::tds::stream::token Login failed for user...)` error that the login failed and a `RuntimeError('Timed out in bb8')` error:

```python
from urllib import parse
import connectorx as cx

u = parse.quote_plus("FEHCH\oper_dev")
p = parse.quote_plus("b3k@!!1aa")
host = "fehchdev.com"
db = "registrar_home"
uri = f"mssql://{u}:{p}@{host}:1433/{db}"

cx.read_sql(uri, "select * from dbs1.users_home")
```

> RuntimeError('Timed out in bb8')

I've tried several combinations too, the only difference is `encrypt=true` takes longer to log any errors:

```python
uri = f"mssql://{u}:{p}@{host}:1433/{db}?trusted_connection=true&encrypt=true"
uri = f"mssql://{u}:{p}@{host}:1433/{db}?trusted_connection=true&encrypt=false"
uri = f"mssql://{u}:{p}@{host}:1433/{db}?trusted_connection=false&encrypt=false"
uri = f"mssql://{u}:{p}@{host}:1433/{db}?driver=SQL+Server&trusted_connection=true&encrypt=true"

```

**Reply by dataengineergaurav** [reply]
facing the same issue.
